docs/decisions/0011-producing-to-event-bus-via-settings.rst

```diff
@@ -0,0 +1,92 @@
+
+The current implementation of openedx-events does not actually push any events to the underlying implementations like `edx-event-bus-kafka`_ and `edx-event-bus-redis`_. The event-producing application is expected to create a signal handler (since openedx-events subclasses Django signals) to catch the event and push it into the event bus. Some examples of the handlers: `handlers example`_.
+
+.. _handlers example: https://github.com/openedx/edx-platform/blob/27b8d2f68d5dfaf84755e7d7f8dccc97ce3be509/cms/djangoapps/contentstore/signals/handlers.py#L162-L210
+.. _edx-event-bus-kafka: https://github.com/openedx/event-bus-kafka
+.. _edx-event-bus-redis: https://github.com/openedx/event-bus-redis
+
+This ADR aims to propose a solution for configuring the details, like the topic name, consumer group name, etc. via Django settings as well as pushing events to the event bus without requiring the producing application to write additional handlers.
+
+
+Decision
+********
+
+Update constructor of ``OpenEdxPublicSignal`` class to optionally accept ``topics`` and ``event_key_field`` and store them. These values can be then used as default while pushing them to event bus implementations. Add ``topics`` and ``event_key_field`` for existing signal definitions in openedx_events repository. Once this ADR is implemented, make ``topics`` and ``event_key_field`` required in ``OpenEdxPublicSignal`` constructor.
+
+Create a generic signal handler to push events to the event bus. This handler should be attached to or connected to the signals that are enabled in Django settings. The configuration format will be as shown below:
+
+.. code-block:: python
+
+   # .. setting_name: EVENT_BUS_PRODUCER_CONFIG
+   # .. setting_default: By default all signals are configured using initialization values from signal but not enabled.
+   # .. setting_description: Dictionary specifying event_types (keys) and its configuration dictionary as values.
+   #   Each configuration dictionary contains a flag denoting whether the event will be enabled or disabled and
+   #   optionally details like topics/stream names where the event will be pushed to, event_key_field which is
+   #   and period delimited string path to event data field to use as event key.
+   #   Note: The topic names should not include environment prefix as it will be dynamically added based on
+   #   EVENT_BUS_TOPIC_PREFIX setting.
+   EVENT_BUS_PRODUCER_CONFIG = {
+       'org.openedx.content_authoring.xblock.deleted.v1': {
+           'event_key_field': 'xblock_info.usage_key', # override default event_key_field
+           'enabled': True,
```

Comment:
Given that this `enabled` boolean would be inside a dictionary structure, we wouldn't be able to (readily) override it separately from the rest of the `EVENT_BUS_PRODUCER_CONFIG`. I'm not sure how much value it adds over e.g. commenting out the unwanted entries.

(In contrast, disabling *all* event production could be done by setting the producer callable reference to an empty string, so that no implementation is loaded.)

Reply:
@timmc-edx I am not sure if I follow the issue here. Could you please explain with an example.

Reply:
I was actually thinking that the default configs would exist, but that they'd be disabled at first. I've really grown to not like commented out code (which I could explain more). This code feels more legitimate, even though it wouldn't actually do anything (for now) and you'd have to literally copy it to enable it. That said, in the future, by default we could simply enable events that the platform really is depending on.